_posts/2024-01-25-websockets-in-http4s.md

```diff
@@ -0,0 +1,1561 @@
+
+In case the user doesn't belong to a room, we return `Seq.empty[OutputMessage].pure[F]`, an empty sequence, however, if he or she is already in a room, we first update the members by getting all room members and removing the user by calling `cs.roomMembers.getOrElse(room, Set()) - user` then proceed to update the `stateRef`:
+```scala
+stateRef
+  .update(ccs =>
+    ChatState(
+      ccs.userRooms - user,
+      if (updateMembers.isEmpty) {
+        ccs.roomMembers - room
+      } else {
+        ccs.roomMembers + (room -> updateMembers)
+      }
+    )
+  )
+```
+
+First, we call `ccs.userRooms - user` to remove the user from `userRooms`, then we check if `updateMembers` is empty, if this is true we call `ccs.roomMembers - room` to remove the room from `roomMembers` since we don't want idle rooms with no users. Otherwise, we call `ccs.roomMembers + (room -> updateMembers)` to update the room members:
+
+```scala
+object Command {
+  def make[F[_]: Monad](chatState: Ref[F, ChatState]): Command[F] =
+    new Command[F] {
+      ...
+      override def disconnect(
+          userRef: Ref[F, Option[User]]
+      ): F[Seq[OutputMessage]] =
+        userRef.get.flatMap {
+          case Some(user) => removeFromCurrentRoom(chatState, user)
+          case None       => Seq.empty[OutputMessage].pure[F]
+        }
+      ...
+    }
+}
+```
+Finally the `disconnect()` function removes the user from the current room in case the WebSocket connection is broken or cut to reflect an up-to-date `ChatState`.
+
+### 6.4. Parsing String Inputs
+
+In this section, we'll learn how to parse the input messages we receive from the WebSocket. Create `InputMessage.scala` in `src/main/scala/rockthejvm/websockets/InputMessage.scala` and add the follwing code:
+
+```scala
+package rockthejvm.websockets
+
+import cats.data.Validated
+import cats.effect.kernel.Ref
+import cats.Monad
+import cats.syntax.all.*
+
+trait InputMessage[F[_]] {
+  def defaultRoom: Validated[String, Room]
+  def parse(
+      userRef: Ref[F, Option[User]],
+      text: String
+  ): F[Seq[OutputMessage]]
+}
+
+object InputMessage {
+  def make[F[_]: Applicative: FlatMap](
+      command: Command[F]
+  ): InputMessage[F] =
+    new InputMessage[F] {
+      override def parse(
+          userRef: Ref[F, Option[User]],
+          text: String
+      ): F[OutputMessage] = ???
+    }
+}
+```
+
+We start by creating an `InputMessage` trait with `defaultRoom`, a function that returns `Validated[String, Room]` and a `parse()` function that takes a `Ref[F, Option[User]]` and a `String` and returns an `F[Seq[OutputMessage]]`. 
+
+Similar to `Command` we create a `make()` where we provide the implementations for `InputMessage`. We make sure to bind `Monad` to `F[_]` for the following implementation of `parse()`.
+
+```scala
+...
+override def parse(
+    userRef: Ref[F, Option[User]],
+    text: String
+): F[OutputMessage] =
+  text.trim match
+    case "" => DiscardMessage.pure[F]
+    case txt @ _ =>
```

Comment:
`case txt` is enough

Reply:
This part is needed, without it, `text.charAt(0) == '/'` in `processText4Reg` and `processText4UnReg` would fail

Reply:
Fixed